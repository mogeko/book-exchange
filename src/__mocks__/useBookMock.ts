import * as hooks from "@/lib/hooks/useBooks";
import { exampleData as exampleBooksData } from "@/__mocks__/useBooksMock";
import { genExampleRes } from "@/__mocks__/useQueryMock";
import { faker } from "@faker-js/faker";

const mock = jest.spyOn(hooks, "useBook");

const useBookMock = {
  target: hooks.useBook,
  returnResult: () => ({
    success: () =>
      mock.mockImplementation(() => genExampleRes({ data: exampleData })),
    error: () =>
      mock.mockImplementation(() => genExampleRes({ isError: true })),
    loading: () =>
      mock.mockImplementation(() => genExampleRes({ isLoading: true })),
  }),
};

export const exampleData: BookType = {
  ...exampleBooksData,
  mate: {
    ...exampleBooksData.mate,
    publisher: faker.company.companyName(),
    subtitle: faker.lorem.sentence(10),
    language: "English",
    publication_date: faker.date.past().toISOString(),
    isbn: `978-${faker.datatype.number({ min: 1000000000, max: 1999999999 })}`,
    paperback: faker.datatype.number({ min: 100, max: 1000 }),
  },
  desc: {
    text: faker.lorem.paragraph(50),
    is_folded: faker.datatype.boolean(),
  },
  digest: {
    text: faker.lorem.paragraph(50),
    is_folded: faker.datatype.boolean(),
  },
};

type BookType = hooks.BookType;

export default useBookMock;
