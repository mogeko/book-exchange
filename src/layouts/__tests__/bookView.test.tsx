import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import useBookMock, { exampleData } from "@/__mocks__/useBookMock";
import useRouterMock from "@/__mocks__/useRouter";
import BookView, { BookDesc, BookInfo } from "@/layouts/bookView";

describe("BookView", () => {
  beforeEach(() => {
    useBookMock.returnResult().success();
    useRouterMock.returnResult({ query: { id: "bk1000" } });
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it("render a BookView", () => {
    const { container } = render(<BookView />);

    expect(useRouterMock.target).toBeCalledWith();
    expect(useBookMock.target).toBeCalledWith("bk1000");
    expect(container.querySelector("h1")?.textContent).toBe(exampleData.title);
    expect(container.querySelectorAll("div.inline-flex").length).toBe(
      Object.keys(exampleData.mate).length
    );
  });

  it("snapshot a BookView", () => {
    const { container } = render(<BookView />);

    expect(container).toMatchSnapshot();
  });
});

describe("BookView with abnormal state", () => {
  beforeEach(() => {
    useRouterMock.returnResult({ query: { id: "bk1000" } });
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it("renders a BookView when error occurred", () => {
    useBookMock.returnResult().error();
    const { container } = render(<BookView />);

    expect(container.querySelector("div")?.textContent).toEqual(
      "Network Error!"
    );
  });

  it("snapshot a BookView when error occurred", () => {
    useBookMock.returnResult().error();
    const { container } = render(<BookView />);

    expect(container).toMatchSnapshot();
  });

  it("renders a BookView when loading", () => {
    useBookMock.returnResult().loading();
    const { container } = render(<BookView />);

    expect(container.querySelector(".animate-pulse")).toBeInTheDocument();
  });

  it("snapshot a BookView when loading", () => {
    useBookMock.returnResult().loading();
    const { container } = render(<BookView />);

    expect(container).toMatchSnapshot();
  });
});

describe("BookInfo", () => {
  it("renders a BookInfo without meta data", () => {
    const { container } = render(
      <BookInfo title="title" cover={exampleData.cover} />
    );

    expect(container.querySelector("h1")?.textContent).toBeUndefined();
  });

  it("snapshot a BookInfo", () => {
    const { container } = render(
      <BookInfo title="title" cover={exampleData.cover} />
    );

    expect(container).toMatchSnapshot();
  });
});

describe("BookDesc", () => {
  it("render a BookDesc", () => {
    const { container } = render(
      <BookDesc title="Example Title" text={exampleData.desc?.long_desc} />
    );

    expect(container.querySelector("h2")?.textContent).toBe("Example Title");
    expect(container.querySelector("p")?.textContent).toBe(
      exampleData.desc?.long_desc
    );
  });

  it("render a BookDesc without text", () => {
    const { container } = render(<BookDesc title="Example Title" />);

    expect(container.querySelector("h2")?.textContent).toBeUndefined();
    expect(container.querySelector("p")).toBeNull();
  });

  it("snapshot a BookDesc", () => {
    const { container } = render(
      <BookDesc title="Example Title" text={exampleData.desc?.long_desc} />
    );

    expect(container).toMatchSnapshot();
  });
});
