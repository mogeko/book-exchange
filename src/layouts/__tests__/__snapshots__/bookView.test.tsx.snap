// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookDesc snapshot a BookDesc 1`] = `
<div>
  <div
    class="flex flex-col items-start my-2"
  >
    <h2
      class="text-lg capitalize w-full my-2"
    >
      Example Title
    </h2>
    <p
      class="indent-8 "
    >
      Est sed et pariatur nihil aut vero impedit. Pariatur non sequi voluptas non illo nemo voluptatibus. Amet dolorum soluta culpa itaque esse sint. Excepturi atque quia consequatur excepturi. Doloribus tempora porro illum est aliquid sapiente dolores neque. Et tempore accusamus et porro doloribus. Voluptatibus minus voluptatibus omnis quo. Voluptas unde neque sit sit id aut voluptas maxime. Inventore quasi impedit deleniti nobis sed perferendis repellendus. Occaecati illo voluptates perspiciatis sequi a. Sint tenetur quod. Quod eaque ducimus aperiam mollitia. Ullam voluptas ad temporibus fugiat qui. Corrupti amet at quidem cumque facilis. Perspiciatis quo placeat aspernatur quos. Commodi placeat consequatur sit inventore aspernatur fugiat molestiae distinctio vel. Quasi saepe quo nihil perspiciatis eveniet alias est. Molestiae esse autem iusto quia rem maxime reprehenderit autem. Optio et dolorum autem porro. Aut consequatur pariatur soluta magnam laborum exercitationem eligendi. Dolor corrupti earum rerum non. Vero aperiam animi eum vero aspernatur officiis repellendus commodi. Qui in maxime. Ipsa eaque debitis ab consequatur provident nihil fuga qui. Delectus tenetur sint et et nemo. Deleniti fuga eveniet sit fugiat ea consequatur repellendus. Autem quis fugit est nihil itaque inventore aut. Est placeat delectus veritatis veritatis ipsa voluptatum ut ratione. Aut rerum vel sit blanditiis voluptatum autem occaecati ratione animi. Iure odio delectus veniam neque tenetur natus. Similique quo aperiam beatae natus sequi accusamus. Ea autem sapiente quod ex. Qui sit qui eum qui exercitationem autem iure. Eligendi cupiditate sed. Deserunt quia consequuntur aut placeat sequi. Nulla voluptatem unde occaecati. Eum molestiae a sunt labore ea corporis. Et sed ullam tempora. Et officiis aut magni blanditiis eveniet doloribus. Excepturi autem dolorem tenetur reprehenderit. Excepturi qui voluptatum delectus. Nihil quidem hic est temporibus provident cumque est. Totam voluptas et in dolores non. Et id ea minima repellendus quia sed porro. Suscipit aut accusamus minus est ea iste vitae et eaque. Sed quidem modi id. Aperiam qui quis minus rerum ut. Esse nihil aut possimus autem. Illo rerum et quia molestiae repudiandae non. Numquam alias quos provident quis. Molestiae et commodi ullam occaecati omnis. Magni iusto autem consequatur reiciendis voluptatibus doloremque ut qui iusto. Tenetur atque sed.
    </p>
  </div>
</div>
`;

exports[`BookInfo snapshot a BookInfo 1`] = `
<div>
  <div
    class="my-4 flex w-full gap-3"
  >
    <figure
      class="flex items-center justify-center"
    >
      <div
        class="relative w-32 h-48"
      >
        <span
          style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
        >
          <img
            alt="title"
            data-nimg="fill"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
          <noscript />
        </span>
      </div>
    </figure>
    <div
      class="flex flex-col"
    />
  </div>
</div>
`;

exports[`BookView snapshot a BookView 1`] = `
<div>
  <div
    class="flex flex-col"
  >
    <h1
      class="text-2xl w-full py-2"
    >
      catch
    </h1>
    <div
      class="my-4 flex w-full gap-3"
    >
      <figure
        class="flex items-center justify-center"
      >
        <div
          class="relative w-32 h-48"
        >
          <span
            style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
          >
            <img
              alt="catch"
              data-nimg="fill"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
            <noscript />
          </span>
        </div>
      </figure>
      <div
        class="flex flex-col"
      >
        <div
          class="inline-flex gap-1 text-base"
        >
          <span
            class="capitalize"
          >
            author
            :
          </span>
          <span>
            Gracie
          </span>
        </div>
        <div
          class="inline-flex gap-1 text-base"
        >
          <span
            class="capitalize"
          >
            publisher
            :
          </span>
          <span>
            Cremin, Grady and Yost
          </span>
        </div>
        <div
          class="inline-flex gap-1 text-base"
        >
          <span
            class="capitalize"
          >
            subtitle
            :
          </span>
          <span>
            Aut laborum perferendis harum quo minus quis consequatur natus nihil.
          </span>
        </div>
        <div
          class="inline-flex gap-1 text-base"
        >
          <span
            class="capitalize"
          >
            language
            :
          </span>
          <span>
            English
          </span>
        </div>
        <div
          class="inline-flex gap-1 text-base"
        >
          <span
            class="capitalize"
          >
            publication_date
            :
          </span>
          <span>
            2022
          </span>
        </div>
        <div
          class="inline-flex gap-1 text-base"
        >
          <span
            class="uppercase"
          >
            isbn
            :
          </span>
          <span>
            978-1430433573
          </span>
        </div>
        <div
          class="inline-flex gap-1 text-base"
        >
          <span
            class="capitalize"
          >
            paperback
            :
          </span>
          <span>
            468
             pages
          </span>
        </div>
      </div>
    </div>
    <div
      class="flex flex-col items-start my-2"
    >
      <h2
        class="text-lg capitalize w-full my-2"
      >
        About this book
      </h2>
      <p
        class="indent-8 "
      >
        Est sed et pariatur nihil aut vero impedit. Pariatur non sequi voluptas non illo nemo voluptatibus. Amet dolorum soluta culpa itaque esse sint. Excepturi atque quia consequatur excepturi. Doloribus tempora porro illum est aliquid sapiente dolores neque. Et tempore accusamus et porro doloribus. Voluptatibus minus voluptatibus omnis quo. Voluptas unde neque sit sit id aut voluptas maxime. Inventore quasi impedit deleniti nobis sed perferendis repellendus. Occaecati illo voluptates perspiciatis sequi a. Sint tenetur quod. Quod eaque ducimus aperiam mollitia. Ullam voluptas ad temporibus fugiat qui. Corrupti amet at quidem cumque facilis. Perspiciatis quo placeat aspernatur quos. Commodi placeat consequatur sit inventore aspernatur fugiat molestiae distinctio vel. Quasi saepe quo nihil perspiciatis eveniet alias est. Molestiae esse autem iusto quia rem maxime reprehenderit autem. Optio et dolorum autem porro. Aut consequatur pariatur soluta magnam laborum exercitationem eligendi. Dolor corrupti earum rerum non. Vero aperiam animi eum vero aspernatur officiis repellendus commodi. Qui in maxime. Ipsa eaque debitis ab consequatur provident nihil fuga qui. Delectus tenetur sint et et nemo. Deleniti fuga eveniet sit fugiat ea consequatur repellendus. Autem quis fugit est nihil itaque inventore aut. Est placeat delectus veritatis veritatis ipsa voluptatum ut ratione. Aut rerum vel sit blanditiis voluptatum autem occaecati ratione animi. Iure odio delectus veniam neque tenetur natus. Similique quo aperiam beatae natus sequi accusamus. Ea autem sapiente quod ex. Qui sit qui eum qui exercitationem autem iure. Eligendi cupiditate sed. Deserunt quia consequuntur aut placeat sequi. Nulla voluptatem unde occaecati. Eum molestiae a sunt labore ea corporis. Et sed ullam tempora. Et officiis aut magni blanditiis eveniet doloribus. Excepturi autem dolorem tenetur reprehenderit. Excepturi qui voluptatum delectus. Nihil quidem hic est temporibus provident cumque est. Totam voluptas et in dolores non. Et id ea minima repellendus quia sed porro. Suscipit aut accusamus minus est ea iste vitae et eaque. Sed quidem modi id. Aperiam qui quis minus rerum ut. Esse nihil aut possimus autem. Illo rerum et quia molestiae repudiandae non. Numquam alias quos provident quis. Molestiae et commodi ullam occaecati omnis. Magni iusto autem consequatur reiciendis voluptatibus doloremque ut qui iusto. Tenetur atque sed.
      </p>
    </div>
    <div
      class="flex flex-col items-start my-2"
    >
      <h2
        class="text-lg capitalize w-full my-2"
      >
        Popular Highlights in this book
      </h2>
      <p
        class="indent-8 "
      >
        Aut earum harum a nulla impedit. Magni fuga et cumque excepturi enim voluptas. Quasi ipsum ut. Quia exercitationem sequi voluptates delectus ullam laborum. Porro et consequuntur harum. Hic praesentium maiores qui asperiores qui vero accusantium nemo voluptas. Sequi voluptatum perferendis aperiam qui assumenda velit necessitatibus. Adipisci animi molestias aspernatur quas vel consequatur veniam. Nisi voluptas eos. Facere laboriosam rerum quidem possimus nisi sit totam veniam. Repellendus magni et similique consequatur vel officiis ullam. Est quidem ullam eum voluptas porro et voluptates doloribus. Aut minima reprehenderit commodi nemo ducimus atque. Minima neque natus consequatur pariatur odit quaerat. Facere voluptatem quidem quia omnis et. Neque cum eveniet magnam vitae sint. Fugit quae consequatur quis quibusdam dolor. Quos odio quam nihil rerum quia assumenda sequi perspiciatis exercitationem. Reprehenderit quia molestiae aut quia voluptatem. Voluptatum dolore officiis facere ea hic nihil voluptatem ut. Consequuntur accusamus similique odit. Repellat culpa explicabo. In ex harum veniam. Corporis fuga saepe facilis. Ut labore similique molestias nihil. Aspernatur reprehenderit neque natus. Consequatur blanditiis odio ad blanditiis ab. Qui qui odit sit dignissimos voluptas veniam. Hic placeat corrupti natus qui eius magni autem iusto. Libero fugit consequatur voluptatibus nisi voluptas enim hic voluptate. Aliquam impedit consectetur. Praesentium libero quisquam. In eligendi labore. Numquam quis voluptas. Aut aut perferendis laborum autem. Sunt et dolore reprehenderit rerum dolorem itaque. In modi minima. Quisquam velit asperiores nemo ipsam iste amet at cumque rerum. Velit est id autem soluta deleniti voluptas. Aut libero nihil iusto deserunt qui id eveniet officia totam. Minima quasi nihil itaque quos assumenda et. Impedit rerum enim id. At fugit alias aliquid quidem enim. Ab repellendus sunt suscipit adipisci nulla odit. Est aut exercitationem non quia ratione itaque molestias qui libero. Odio corporis in. Aut ipsam soluta minus vel vel facere autem modi quos. Eaque voluptas ea quibusdam totam dolor officiis et. Est corporis delectus voluptas est non fugit occaecati numquam dolor. Ipsa consequatur et. Laboriosam nulla eveniet voluptatem fugiat tempore veritatis vitae cumque dolorem. Est tempora itaque cumque nesciunt unde.
      </p>
    </div>
  </div>
</div>
`;

exports[`BookView with abnormal state snapshot a BookView when error occurred 1`] = `
<div>
  <div
    class="alert alert-error shadow-lg"
  >
    <div>
      <svg
        class="w-6 h-6"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M8.6 1c1.6.1 3.1.9 4.2 2 1.3 1.4 2 3.1 2 5.1 0 1.6-.6 3.1-1.6 4.4-1 1.2-2.4 2.1-4 2.4-1.6.3-3.2.1-4.6-.7-1.4-.8-2.5-2-3.1-3.5C.9 9.2.8 7.5 1.3 6c.5-1.6 1.4-2.9 2.8-3.8C5.4 1.3 7 .9 8.6 1zm.5 12.9c1.3-.3 2.5-1 3.4-2.1.8-1.1 1.3-2.4 1.2-3.8 0-1.6-.6-3.2-1.7-4.3-1-1-2.2-1.6-3.6-1.7-1.3-.1-2.7.2-3.8 1-1.1.8-1.9 1.9-2.3 3.3-.4 1.3-.4 2.7.2 4 .6 1.3 1.5 2.3 2.7 3 1.2.7 2.6.9 3.9.6zM7.9 7.5L10.3 5l.7.7-2.4 2.5 2.4 2.5-.7.7-2.4-2.5-2.4 2.5-.7-.7 2.4-2.5-2.4-2.5.7-.7 2.4 2.5z"
          fill-rule="evenodd"
        />
      </svg>
      <span>
        Network Error!
      </span>
    </div>
  </div>
</div>
`;

exports[`BookView with abnormal state snapshot a BookView when loading 1`] = `
<div>
  <div
    class="flex flex-col"
  >
    <h1
      class="text-2xl w-full py-2"
    />
    <div
      class="animate-pulse"
    >
      <div
        class="my-4 flex w-full gap-3"
      >
        <div
          class="relative w-32 h-48 bg-slate-700 rounded"
        />
        <div
          class="flex flex-col gap-3"
        >
          <div
            class="inline-flex gap-2"
          >
            <div
              class="w-14 h-2 bg-slate-700 rounded"
            />
            <div
              class="w-32 h-2 bg-slate-700 rounded"
            />
          </div>
          <div
            class="inline-flex gap-2"
          >
            <div
              class="w-14 h-2 bg-slate-700 rounded"
            />
            <div
              class="w-32 h-2 bg-slate-700 rounded"
            />
          </div>
          <div
            class="inline-flex gap-2"
          >
            <div
              class="w-14 h-2 bg-slate-700 rounded"
            />
            <div
              class="w-32 h-2 bg-slate-700 rounded"
            />
          </div>
          <div
            class="inline-flex gap-2"
          >
            <div
              class="w-14 h-2 bg-slate-700 rounded"
            />
            <div
              class="w-32 h-2 bg-slate-700 rounded"
            />
          </div>
          <div
            class="inline-flex gap-2"
          >
            <div
              class="w-14 h-2 bg-slate-700 rounded"
            />
            <div
              class="w-32 h-2 bg-slate-700 rounded"
            />
          </div>
          <div
            class="inline-flex gap-2"
          >
            <div
              class="w-14 h-2 bg-slate-700 rounded"
            />
            <div
              class="w-32 h-2 bg-slate-700 rounded"
            />
          </div>
          <div
            class="inline-flex gap-2"
          >
            <div
              class="w-14 h-2 bg-slate-700 rounded"
            />
            <div
              class="w-32 h-2 bg-slate-700 rounded"
            />
          </div>
          <div
            class="inline-flex gap-2"
          >
            <div
              class="w-14 h-2 bg-slate-700 rounded"
            />
            <div
              class="w-32 h-2 bg-slate-700 rounded"
            />
          </div>
          <div
            class="inline-flex gap-2"
          >
            <div
              class="w-14 h-2 bg-slate-700 rounded"
            />
            <div
              class="w-32 h-2 bg-slate-700 rounded"
            />
          </div>
          <div
            class="inline-flex gap-2"
          >
            <div
              class="w-14 h-2 bg-slate-700 rounded"
            />
            <div
              class="w-32 h-2 bg-slate-700 rounded"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="animate-pulse"
    >
      <div
        class="flex flex-col my-4 gap-3"
      >
        <div
          class="h-4 w-44 mb-4 bg-slate-700 rounded"
        />
        <div
          class="inline-flex"
        >
          <div
            class="w-8"
          />
          <div
            class="w-full h-2 bg-slate-700 rounded"
          />
        </div>
        <div
          class="w-full h-2 bg-slate-700 rounded"
        />
        <div
          class="w-full h-2 bg-slate-700 rounded"
        />
        <div
          class="w-full h-2 bg-slate-700 rounded"
        />
        <div
          class="w-full h-2 bg-slate-700 rounded"
        />
        <div
          class="w-64 h-2 bg-slate-700 rounded"
        />
      </div>
    </div>
    <div
      class="animate-pulse"
    >
      <div
        class="flex flex-col my-4 gap-3"
      >
        <div
          class="h-4 w-44 mb-4 bg-slate-700 rounded"
        />
        <div
          class="inline-flex"
        >
          <div
            class="w-8"
          />
          <div
            class="w-full h-2 bg-slate-700 rounded"
          />
        </div>
        <div
          class="w-full h-2 bg-slate-700 rounded"
        />
        <div
          class="w-full h-2 bg-slate-700 rounded"
        />
        <div
          class="w-full h-2 bg-slate-700 rounded"
        />
        <div
          class="w-full h-2 bg-slate-700 rounded"
        />
        <div
          class="w-64 h-2 bg-slate-700 rounded"
        />
      </div>
    </div>
  </div>
</div>
`;
